File: tester/harmony/rnoh/src/main/cpp/RNOHCorePackage/TurboModules/ExceptionsManagerTurboModule.cpp
#include "ExceptionsManagerTurboModule.h"
#include "RNOH/ArkTSTurboModule.h"

using namespace rnoh;
using namespace facebook;

rnoh::ExceptionsManagerTurboModule::ExceptionsManagerTurboModule(const ArkTSTurboModule::Context ctx, const std::string name)
    : rnoh::ArkTSTurboModule(ctx, name) {
    methodMap_ = {
        ARK_METHOD_METADATA(reportFatalException, 3),
        ARK_METHOD_METADATA(reportSoftException, 3),
        ARK_METHOD_METADATA(reportException, 1),
        ARK_METHOD_METADATA(updateExceptionMessage, 3),
        ARK_METHOD_METADATA(dismissRedbox, 0)
    };
}

File: tester/harmony/rnoh/src/main/ets/RNOHCorePackage/turboModules/ExceptionsManagerTurboModule.ts
import { TurboModule } from "../../RNOH/TurboModule";

export type StackFrame = {
  column?: number,
  file?: string,
  lineNumber?: number,
  methodName: string,
  collapse?: boolean,
};

export type ExceptionData = {
  message: string,
  originalMessage?: string,
  name?: string,
  componentStack?: string,
  stack: StackFrame[],
  id: number,
  isFatal: boolean,
  extraData?: Object,
};

export class ExceptionsManagerTurboModule extends TurboModule {
  reportFatalException(message: string, stack: StackFrame[], exceptionId: number): void {
    this.ctx.logger.error(`ExceptionsManager::reportFatalException ${message}`);
    stack.forEach((frame) => {
      this.ctx.logger.error(JSON.stringify(frame));
    });

    throw new Error(message);
  }

  reportSoftException(message: string, stack: StackFrame[], exceptionId: number): void {
    this.ctx.logger.error(`ExceptionsManager::reportSoftException ${message}`);
    stack.forEach((frame) => {
      this.ctx.logger.error(JSON.stringify(frame));
    });
  }

  reportException(data: ExceptionData): void {
    this.ctx.logger.error(`ExceptionsManager::reportException ${data.message}`);
    data.stack.forEach((frame) => {
      this.ctx.logger.error(JSON.stringify(frame));
    });

    if (data.isFatal) {
      throw new Error(data.message);
    }
  }

  updateExceptionMessage(message: string, stack: StackFrame[], exceptionId: number): void {
    this.ctx.logger.error(`ExceptionsManager::updateExceptionMessage ${message}`);
    stack.forEach((frame) => {
      this.ctx.logger.error(JSON.stringify(frame));
    });
  }

  dismissRedbox(): void {
    // NOOP for now
  }
}